<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en-us" lang="en-us"><head><title>TheAnnotator &lt; Ponting &lt; TWiki</title>


<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-15">
<link rel="icon" href="https://wiki.anat.ox.ac.uk/twiki/pub/Ponting/WebPreferences/favicon.ico" type="image/x-icon">
<link rel="shortcut icon" href="https://wiki.anat.ox.ac.uk/twiki/pub/Ponting/WebPreferences/favicon.ico" type="image/x-icon">
<link rel="alternate" type="application/rss+xml" title="RSS Feed" href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebRss">
<!-- base href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/TheAnnotator" -->

<script type="text/javascript" src="TheAnnotator_files/twiki.js"></script>
<script type="text/javascript" src="TheAnnotator_files/pattern.js"></script>
<script type="text/javascript">
//<![CDATA[
	function launchWindow(inWeb, inTopic) {
		return launchTheWindow("/twiki/bin/view/", inWeb, inTopic, "", "viewplain" );
	}
//]]>
</script>




<style type="text/css" media="all">
	#patternTopBar,
#patternClearHeaderCenter,
#patternClearHeaderLeft,
#patternClearHeaderRight,
#patternTopBarContentsOuter {
	height:64px; /* top bar height; make room for header columns */
	overflow:hidden;
}
	#patternOuter {
	margin-left:12em;
}
#patternLeftBar {
	width:12em;
	margin-left:-12em;
}
</style>
<style type="text/css" media="all">
	@import url('/twiki/pub/TWiki/PatternSkin/layout.css');
	@import url('/twiki/pub/TWiki/PatternSkin/style.css');
	@import url('/twiki/pub/TWiki/PatternSkin/colors.css');
</style><!--[if IE]>
<style type="text/css" media="screen">
pre {
	overflow-x:auto;
	padding-bottom:expression(this.scrollWidth > this.offsetWidth ? 16 : 0);
}
</style>
<![endif]--><!--[if lte IE 6]> 
<style type="text/css">
#patternLeftBar {
	position:relative; /* IE 5.5 needs this or the contents won't show outside the parent container on print. IE 6.0 needs it only during printable copy! */
}
</style>
<![endif]-->



<style type="text/css" media="all">
	/* Styles that are set using variables */
	#patternLeftBar .patternWebIndicator,
	.patternBookView .twikiTopRow {
		background-color:#CCFFCC;
	}
	#patternTopBar { background-image:url("/twiki/pub/TWiki/PatternSkin/TWiki_header.gif"); background-repeat:no-repeat;}
	#patternTopBar { background-color:#ffffff;}
	.patternBookView {
		border-color:#CCFFCC;
	}
	.patternPreviewPage #patternMain {
		/* uncomment to set the preview image */
		/*background-image:url("/twiki/pub/TWiki/PreviewBackground/preview2bg.gif");*/
	}
	
	.twikiTable .tableSortIcon img {visibility:hidden;}
	.twikiTable .tableSortUp {background:url("https://wiki.anat.ox.ac.uk/twiki/pub/TWiki/TWikiDocGraphics/tablesortup.gif") no-repeat;}
	.twikiTable .tableSortDown {background:url("https://wiki.anat.ox.ac.uk/twiki/pub/TWiki/TWikiDocGraphics/tablesortdown.gif") no-repeat;}
	.twikiTable .tableSortDiamond {background:url("https://wiki.anat.ox.ac.uk/twiki/pub/TWiki/TWikiDocGraphics/tablesortdiamond.gif") no-repeat;}
</style>
<style type="text/css" media="all">
	
	
	
</style>


<style type="text/css" media="all">
	@import url("/twiki/pub/TWiki/PatternSkin/print.css");
</style></head><body class="patternViewPage"><a name="PageTop"></a>
<div id="patternScreen">
<div id="patternPageShadow">
<div id="patternPage">
<div id="patternWrapper"><div id="patternOuter">
<div id="patternFloatWrap">
<div id="patternMain"><div id="patternClearHeaderCenter"></div>
<div id="patternMainContents">
<div class="patternTop">
<div class="patternToolBar">
<div class="patternToolBarButtons"><span class="patternButton"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/edit/Ponting/TheAnnotator?t=1191322629" rel="nofollow" title="Edit this topic text" accesskey="e"><span class="twikiAccessKey">E</span>dit</a></span><span class="patternButton"><span class="patternButton"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/attach/Ponting/TheAnnotator" rel="nofollow" title="Attach an image or document to this topic" accesskey="a"><span class="twikiAccessKey">A</span>ttach</a></span></span><span class="patternButton"><span class="patternButton"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/TheAnnotator?cover=print" rel="nofollow" title="Printable version of this topic" accesskey="p"><span class="twikiAccessKey">P</span>rintable</a></span></span>
</div>
</div><!-- /patternToolBar-->
<div class="patternToolBarBottom"></div>
</div><!-- /patternTop-->
<div class="patternHomePath"><span class="twikiRight"> <span class="patternRevInfo">r9 - 02 Oct 2007 - 10:57:08 - <a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Main/GertonLunter" class="twikiLink">GertonLunter</a></span></span><span class="patternHomePathTitle">You are here: </span> <a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Main/WebHome" class="twikiLink">TWiki</a><span class="twikiSeparator">&nbsp;&gt;&nbsp;</span><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebHome"><img src="TheAnnotator_files/web-bg-small.gif" alt="" style="background-color: rgb(204, 255, 204);" border="0" height="13" width="13">&nbsp;Ponting Web</a><span class="twikiSeparator">&nbsp;&gt;&nbsp;</span><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/GertonLunter" class="twikiLink">GertonLunter</a><span class="twikiSeparator">&nbsp;&gt;&nbsp;</span><span title="Ponting.TheAnnotator">TheAnnotator</span><br class="twikiClear"></div><!-- /patternHomePath--><div class="twikiBeforeText"></div><div class="patternContent"><div class="patternTopic"> -- <a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Main/GertonLunter" class="twikiLink">GertonLunter</a> - 19 Jan 2006
<p>
</p><h3><a name="The_Annotator"></a> The Annotator </h3>
<p>
</p><h2><a name="Todo"></a> Todo </h2>
<p> </p><ul>
<li> New input format, more GFF-like, or straight from mySQL tables
</li> <li> Allow overlapping segments and weigh them?  (This would require some work...)
</li> <li> Better sampling algorithm (allow segments to extend across workspace boundaries)
</li> <li> Writeup...
</li></ul> 
<p>
</p><h2><a name="What_and_why"></a> What and why </h2>
<p>
The Annotator is a Java program to analyze functional overrepresentation of 'things of interest' 
in the genome.  A 'thing' here is anything that can be represented by sets of segments in the
genome (any genome, really).
</p><p>
Existing tools, such as <a href="http://david.niaid.nih.gov/david/" target="_top">Ease</a>, exist that analyze functional 
overrepresentations in gene lists, but I could find no tools that would analyze the overrepresentation
at the nucleotide level, instead of the gene level, i.e. that take into account the number of nucleotides,
instead of just the number of genes.
</p><p>
GONOME, by John Mattick et al (<a href="http://www.pubmedcentral.nih.gov/articlerender.fcgi?tool=pubmed&amp;pubmedid=16504139" target="_top">PMID 16504139</a>)
does something similar (but cannot account for GC biases, and is gene-centric).
</p><p>
</p><h2><a name="What_does_it_do_overview"></a> What does it do, overview </h2>
<p>
Input to the program are 3 sets of regions or DNA positions: </p><ul>
<li> A list of segments of interest
</li> <li> A "workspace"
</li> <li> Annotations
</li></ul> 
The Annotator answers the question: are the segments of interest over- or under-represented
within the annotated regions, compared to the expected coverage if the segments were randomly 
(i.e. uniformly) distributed over the genome.  The Annotator only considers annotations and segments 
that fall within a user-defined 'workspace'.  This allows you to consider only segments that fall e.g. 
in coding regions, intronic regions, or intergenic regions.
<p>
The workspace is allowed to be composed of a number of non-overlapping 'isochores', which allows
to control for confounding variables like C+G content.  (Note that in The Annotator's terminology,
an isochore is a subset of a particular chromosome, not a subset of the whole genome.)
</p><p>
If you're interested in densities for particular chromsomes, or for particular isochores, or 
when you look at everything except the highest GC bits: it is all possible; you can specify 
arbitrary combinations of isochores (which are located at specific chromosomes) and analyze
things there. 
</p><p>
</p><h2><a name="The_algorithm"></a> The algorithm </h2>
<p>
To simplify notation, assume there are no isochores, and just one chromosome.
Let S be the set of segments of interest, W the workspace, and A the regions that carry
a specific annotation.
</p><p>
First, the "observed proportion" of segments that carry the annotation A is computed:
</p><p>
prop(observed) = # (S &#8745; W &#8745; A) / # (S &#8745; W)
</p><p>
Here &#8745; denotes set intersection, and # is the "size-of" operator (the cardinality of the set).
The resulting ratio relates to annotated segments, as a proportion of segments that fall within the workspace.
</p><p>
Next, segments are sampled "randomly" (i.e. uniformly) according to a sampling procedure.
The length distribution of segments follows the empirical length distribution of those
segments in S that have overlap with the workspace.  (More precisely, the length of any segment 
that has nonzero overlap with W is included in the length distribution; as an option, the truncated
length may be used - see below).  After drawing a length
from the distribution, a genomic location is drawn uniformly.  If the resulting segment has any overlap
with the workspace, it is kept, otherwise a new location is drawn.  Before adding the segment to
the sampled set, it is truncated to the workspace. 
The procedure is continued until the number
of nucleotides within the sampled set is (exactly) equal to the number of nucleotides in the intersection
W &#8745; S.  
</p><p>
Some details: Nucleotides that are covered by more than one sampled segment are not counted twice.
The final segment may cause the set to contain more nucleotides than required; in this case the
final segment is truncated.  Secondly, the default behaviour of not truncating segments from S when 
calculating the empirical segment length distribution can be changed.  What is the best choice
depends on the data.  Not truncating can lead to a biased length distribution if it is expected
that segments that only partially overlap the workspace have very different lengths.  However,
truncating can lead to spurious short segments.
</p><p>
Say T is the set of sampled segments, then # (T &#8745; W &#8745; A) / # (W &#8745; A) is computed (the "sampled proportion"),
and it is noted how often this exceeds or falls short of the observed proportion.  These numbers,
divided by the total number of iterations, gives the p value for under- respectively over-representation
of segments with respect to the annotation.
</p><p>
Now for isochores.  These are partitionings of a chromosome into non-overlapping regions, for 
instance, isochores.  When these are defined, the procedure stays the same, except that both
nucleotide counts and length distributions are computed separately for each isochore separately.
More precisely, the procedure is carried out with W replaced by W' = W &#8745; I, where I is the isochore.
This accounts for any density- or length-dependence of segments with respect to isochores.
</p><p>
</p><h2><a name="Input_format"></a> Input format </h2>
<p>
The input consists of up to six files: a segment file, a workspace file, an optional synonym file, an identifier file,
an annotation file, and an optional subset file.  The identifier and annotation files together define the annotations, 
while the synonym file places isochores on chromosomes.
</p><p>
The files are line-oriented, and lines relevant to The Annotator always start with two <code>#</code> signs.
All other lines are ignored.
</p><p>
The program does actually not care which data you put in what file; you can put everything in a single
big file.  This is sometimes useful for things that really belong together, such as annotations and identifiers.
</p><p>
</p><h5><a name="Segment_file"></a><a name="Segment_file_"></a> Segment file: </h5>
<p>
<code>##Seg&nbsp; chromosome1&nbsp; segment1&nbsp; segment2 &nbsp; ... &nbsp; segmentN</code>
</p><p>
<code>##Seg&nbsp; chromosome2&nbsp; segment1 ...</code>
</p><p>
<code>...</code>
</p><p>
The labels <code>chromosome1</code>, <code>chromosome2</code> may be any alphanumeric label (without whitespace).
The various entries must be separated by tabs; spaces surrounding it may be ignored.
Segments have the format
</p><p>
<code>(start, end)</code>
</p><p>
where <code>start</code> and <code>end</code> denote the first nucleotide position in the segment, and the first nucleotide
position beyond the segment (half-open segment convention).  All segments for a particular chromosome 
should be listed in a single line.  Segments themselves may appear in any order.  If segments overlap,
they will be merged.  If <code>end</code> &lt;= <code>start</code>, the segment is deemed empty.
</p><p>
</p><h5><a name="Workspace_and_synonym_file"></a><a name="Workspace_and_synonym_file_"></a> Workspace and synonym file: </h5>
<p>
<code>##Work <tab> isochore1 <tab> segment1 <tab> segment2 <tab> ... <tab> segmentN</tab></tab></tab></tab></tab></code>
</p><p>
<code>...</code>
</p><p>
The format is the same as for segment files.   The segment list may be empty.
The <code>isochore</code> label may again be any alphanumeric label.
</p><p>
Several workspaces may be provided for a single isochore; this is usually done using the
options <code>-workspace</code>, <code>-workspace2</code> and <code>-workspace3</code>.  These workspaces will be intersected
before the algorithm starts.  It is permitted to define workspaces for the same isochore
multiple times in one input file; again, only those sites in the intersection will be used.
</p><p>
To link segments and annotations on the one hand with workspace on the other, the isochores
must be put on chromosomes.  This is done with <code>##Synonym</code> entries:
</p><p>
<code>##Synonym <tab> isochore <tab> chromosome</tab></tab></code>
</p><p>
These may be put in a separate file, but may also appear in any other file.
</p><p>
If the isochore entires in the <code>##Work</code> lines coincide with chromosome labels,
this implicitly makes the isochore coincide with the entire chromosome, and the
synonym line is not necessary.
</p><p>
</p><h5><a name="Annotation_and_Identifier_files"></a> Annotation and Identifier files </h5>
<p>
Annotations are defined in terms of identifiers, which may be throught of as corresponding
to locuslink numbers referring to genes on the genome.
</p><p>
<code>##Id <tab> id <tab> chromosome <tab> segment</tab></tab></tab></code>
</p><p>
The <code>id</code> identifier must be an integer.  Usually, an identifier is associated to just
one segment (e.g. a gene transcript), but more than one segment is allowed (e.g. all LINE
elements on a chromosome).  An identifier cannot be associated to more than one chromosome
(although it can straddle multiple isochores).
There is no requirement for ids not to overlap, or to avoid isochore boundaries.
</p><p>
Annotations are defined as collections of ids:
</p><p>
<code>##Ann <tab> annotation <tab> id1 <tab> id2 <tab> ... <tab> idN</tab></tab></tab></tab></tab></code>
</p><p>
The <code>annotation</code> keyword identifies the annotation.  It may contain whitespace,
and must be followed by a tab.  
</p><p>
</p><h5><a name="Subset_files"></a> Subset files </h5>
<p>
By default, the Annotator looks at densities in individual isochores, but lumps all results together
to report an overall over- or under-representation.  However you can select subsets of isochores and
restrict the analysis to these (e.g. individual chromosomes, or per-isochore analyses).  This is done
by specifying subsets, as follows:
</p><p>
<code>##Subset <tab> id <tab> isochore</tab></tab></code>
</p><p>
Here <code>id</code> is any identifier to referring to some subset; <code>isochore</code> refers to the isochore that
should be included.  Several lines may share the same <code>id</code> to include several isochores into
one subset.
</p><p>
If no subset files are provided, it is assumed that all isochores should be taken together.
</p><p>
</p><h2><a name="Where_is_the_program"></a> Where is the program </h2>
<p>
The program is located in ~gerton/Progs/Annotator/.  It can be started by the script called Annotator in that
directory.  Source files are in app/ and segment/.  All the executable files (.jar files) are in lib/.
</p><p>
One thing to change in the startup script is the -Xmx100M option, which tells the Java interpreter to use
at most 100 Mb of memory.  Java can easily run out of memory with large input files.
</p><p>
</p><h2><a name="Command_line_options"></a> Command line options </h2>
<p>
 -annotation      Annotation-to-id map (filename)
</p><p>
 -ids             Id-to-segment map (filename)
</p><p>
 -segments        Input data (filename)
</p><p>
 -synonyms        Chromosome synonyms for ids and segments (filename)
</p><p>
 -workspace       Genomic segments to consider (filename)
</p><p>
 -workspace2      Genomic segments to consider.  Segments in this workspace will
                  be intersected with those from the other.
</p><p>
 -subset          Subsets to analyze.
</p><p>
 -verbose         Verbosity level (default 1, range 0-5.)
</p><p>
 -mindensity      Annotations with an observed or expected proportion below this number will not be shown.
                  default 0.0001.
</p><p>
 -fdrmaxp         Maximum p-value considered for false-discovery-rate summary (default 0.01).  Decrease to use less memory.
</p><p>
 -baseline        Baseline for over/underrepresentation p-value, relative
                  to randomized proportion (default 1.00).  This is only useful
                  if you want to compare to a different proportion than the
                  actual observed proportion.  All output stays the same except everything
                  related to p values.
</p><p>
 -h               Print help for this application
</p><p>
 -histsize        Histogram size (number of buckets) for empirical segment length distribution
                  (default 100000).  If segments longer than histsize * bucketsize (see below) 
                  are encountered, they are truncated and a warning message is printed.
</p><p>
 -bucketsize      Bucket size for the empirical segment length histogram (default 1).  
</p><p>
 -iterations      Number of iterations (default 100)
</p><p>
 -keepoverhang    Keep overhang for empirical seglen distribution in
                  workspace intersection (default true).  See discussion about empirical length
                  distribution above.
</p><p>
 -maxp            Threshold for p value (default 1.0)  Anything above this p value will not be printed.
</p><p>
 -minz            Threshold for z score (default 0.0)  The z score is approximate, and is computed by comparing
                  the observed proportion with the mean and variance of the expected proportions.  This is useful
                  to estimate p values when they get very low.  Entries where the estimated z score (in absolute value)
                  is below minz are not printed.  I usually set this to 0.0.
</p><p>
 -seed            Random seed (default 1)  
</p><p>
</p><h2><a name="Example_files"></a> Example files </h2>
<p>
Examples are in the Examples/ directory.
</p><p>
id.hg17.txt:  locuslink identifiers, to be used with all annotation files
</p><p>
ann.ease.hg17.txt:  annotations with <span class="twikiNewLink">UniProt<a href="https://wiki.anat.ox.ac.uk/twiki/bin/edit/Ponting/UniProt?topicparent=Ponting.TheAnnotator" rel="nofollow" title="Create this topic">?</a></span>  keywords (data taken from Ease, that's why the name)
</p><p>
ann.expr.gnf.cns.hg17.txt: annotations by tissue-specific expression
(four levels; +++ means 32-fold overexpression compared to median over
all tissues)
</p><p>
ann.gobp.hg17.txt:  annotations using Go BP terms
</p><p>
ann.gocc.hg17.txt:  Go CC terms
</p><p>
ann.gomf.hg17.txt:  Go MF terms
</p><p>
seg.hm.local.hg17.txt:  Indel-purified but fast evolving segments
</p><p>
syn.hg17.cplusg.txt:  Isochore synonym file, to be used with the 'cplusg' workspaces below
</p><p>
work.allassembled.hg17.txt:  All assembled sequence
</p><p>
work.coding.hg17.txt:  All coding exons (according to Ensembl)
</p><p>
work.coding.cplusg.hg17.txt:  All coding exons, and divided into G+C isochores using 10kb non-overlapping windows
</p><p>
work.intron.cplusg.genscan.hg17.txt:  All introns (i.e. Ensembl-annotated genes from first to last exon, minus all
  Ensembl and <span class="twikiNewLink">GenScan<a href="https://wiki.anat.ox.ac.uk/twiki/bin/edit/Ponting/GenScan?topicparent=Ponting.TheAnnotator" rel="nofollow" title="Create this topic">?</a></span>  exons), further divided into G+C isochores.
</p><p>
work.intron.genscan.hg17.txt:  Same without G+C division
</p><p>
</p><p>
</p><h2><a name="Post_processing"></a> Post-processing </h2>
<p>
The program dumps its output to standard out, so you'll have to pipe it to a file.
</p><p>
I set the thresholds so that I get information on all annotations, which can be a pretty big list.
The script prettify.py in the Scripts directory lists the results, ordered by amount of over- or
underrepresention, and filtered for significance by thresholding on p value.  It works like this:
</p><p>
<code>python prettify.py filename p_value_threshold</code>
</p><p>
where <code>filename</code> is the output of the Annotator.</p></div><!-- /patternTopic-->
<div class="twikiAfterText"></div></div><!-- /patternContent-->
<a name="topic-actions"> </a><div class="patternTopicActions"><div class="patternTopicAction"><span class="patternActionButtons"> <span class="patternButton"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/edit/Ponting/TheAnnotator?t=1191322629" rel="nofollow" title="Edit this topic text" accesskey="e"><span class="twikiAccessKey">E</span>dit</a></span><span class="twikiSeparator">&nbsp;|&nbsp;</span><span class="patternButton"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/attach/Ponting/TheAnnotator" rel="nofollow" title="Attach an image or document to this topic" accesskey="a"><span class="twikiAccessKey">A</span>ttach</a></span><span class="twikiSeparator">&nbsp;|&nbsp;</span><span class="patternButton"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/TheAnnotator?cover=print" rel="nofollow" title="Printable version of this topic" accesskey="p"><span class="twikiAccessKey">P</span>rintable</a></span><span class="twikiSeparator">&nbsp;|&nbsp;</span><span class="patternButton"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/TheAnnotator?raw=on" rel="nofollow" title="View raw text without formatting" accesskey="r"><span class="twikiAccessKey">R</span>aw View</a></span><span class="twikiSeparator">&nbsp;|&nbsp;</span>Backlinks: <span class="patternButton"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/oops/Ponting/TheAnnotator?template=backlinksweb" rel="nofollow" title="Search the Ponting Web for topics that link to here" accesskey="b">We<span class="twikiAccessKey">b</span></a></span>, <span class="patternButton"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/oops/Ponting/TheAnnotator?template=backlinksallwebs" rel="nofollow" title="Search all webs for topics that link to here" accesskey="l">A<span class="twikiAccessKey">l</span>l Webs</a></span><span class="twikiSeparator">&nbsp;|&nbsp;</span><span class="patternRevision"><span class="patternButton"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/rdiff/Ponting/TheAnnotator?type=history" rel="nofollow" title="View total topic history" accesskey="h"><span class="twikiAccessKey">H</span>istory</a></span>: r9&nbsp;<a rel="nofollow" href="https://wiki.anat.ox.ac.uk/twiki/bin/rdiff/Ponting/TheAnnotator?rev1=9;rev2=8">&lt;</a>&nbsp;<a rel="nofollow" href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/TheAnnotator?rev=8">r8</a>&nbsp;<a rel="nofollow" href="https://wiki.anat.ox.ac.uk/twiki/bin/rdiff/Ponting/TheAnnotator?rev1=8;rev2=7">&lt;</a>&nbsp;<a rel="nofollow" href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/TheAnnotator?rev=7">r7</a>&nbsp;<a rel="nofollow" href="https://wiki.anat.ox.ac.uk/twiki/bin/rdiff/Ponting/TheAnnotator?rev1=7;rev2=6">&lt;</a>&nbsp;<a rel="nofollow" href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/TheAnnotator?rev=6">r6</a>&nbsp;<a rel="nofollow" href="https://wiki.anat.ox.ac.uk/twiki/bin/rdiff/Ponting/TheAnnotator?rev1=6;rev2=5">&lt;</a>&nbsp;<a rel="nofollow" href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/TheAnnotator?rev=5">r5</a></span><span class="twikiSeparator">&nbsp;|&nbsp;</span><span class="patternButton"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/oops/Ponting/TheAnnotator?template=oopsmore&amp;param1=9&amp;param2=9" rel="nofollow" title="Delete or rename this topic; set parent topic; view and compare revisions" accesskey="m"><span class="twikiAccessKey">M</span>ore topic actions</a></span></span></div></div><!--/patternTopicActions--><div class="patternMoved"></div><!-- /patternMoved-->
</div><!-- /patternMainContents-->
</div><!-- /patternMain--><div id="patternLeftBar"><div id="patternClearHeaderLeft"></div>
<div id="patternLeftBarContents">
<div class="patternWebIndicator"><b><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebHome">Ponting</a></b></div>
<div class="patternLeftBarPersonal">
Hello <a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Main/GertonLunter" class="twikiLink">Gerton Lunter</a>
<div class="patternLeftBarPersonalContent"><ul><li><a href="https://wiki.anat.ox.ac.uk/twiki/bin/edit/Main/GertonLunterLeftBar?templatetopic=TWiki.WebLeftBarPersonalTemplate&amp;topicparent=GertonLunter">Create personal sidebar</a></li></ul></div></div>
<p> </p><ul>
<li> <b><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebHome" class="twikiCurrentWebHomeLink twikiLink"> <img alt="home" src="TheAnnotator_files/home.gif" align="top" border="0" height="16" width="16"> Ponting Web</a></b>
</li> <li> <a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebTopicCreator?parent=TheAnnotator" target="_top"> <img alt="newtopic" src="TheAnnotator_files/newtopic.gif" align="top" border="0" height="16" width="16"> Create New Topic</a>
</li> <li> <a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebTopicList" class="twikiLink"> <img alt="index" src="TheAnnotator_files/index.gif" align="top" border="0" height="16" width="16"> Index</a>
</li> <li> <a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebSearch" class="twikiLink"> <img alt="searchtopic" src="TheAnnotator_files/searchtopic.gif" align="top" border="0" height="16" width="16"> Search</a>
</li> <li> <a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebChanges" class="twikiLink"> <img alt="changes" src="TheAnnotator_files/changes.gif" align="top" border="0" height="16" width="16"> Changes</a>
</li> <li> <a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebNotify" class="twikiLink"> <img alt="notify" src="TheAnnotator_files/notify.gif" align="top" border="0" height="16" width="16"> Notifications</a>
</li> <li> <a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebStatistics" class="twikiLink"> <img alt="statistics" src="TheAnnotator_files/statistics.gif" align="top" border="0" height="16" width="16"> Statistics</a>
</li> <li> <a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebPreferences" class="twikiLink"> <img alt="wrench" src="TheAnnotator_files/wrench.gif" align="top" border="0" height="16" width="16"> Preferences</a>
</li></ul> 
<p>
</p><hr> <ul>
<li> <strong>Webs</strong>
</li> <li> <span style="white-space: nowrap;"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/FGU/WebHome"><img src="TheAnnotator_files/web-bg.gif" alt="" style="background-color: rgb(255, 216, 170);" border="0" height="16" width="16">&nbsp;FGU</a></span>
</li> <li> <span style="white-space: nowrap;"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/FGUJournalPicks/WebHome"><img src="TheAnnotator_files/web-bg.gif" alt="" style="background-color: rgb(204, 255, 204);" border="0" height="16" width="16">&nbsp;FGUJournalPicks</a></span>
</li> <li> <span style="white-space: nowrap;"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Liu/WebHome"><img src="TheAnnotator_files/web-bg.gif" alt="" style="background-color: rgb(0, 153, 102);" border="0" height="16" width="16">&nbsp;Liu</a></span>
</li> <li> <span style="white-space: nowrap;"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Main/WebHome"><img src="TheAnnotator_files/web-bg.gif" alt="" style="background-color: rgb(255, 239, 166);" border="0" height="16" width="16">&nbsp;Main</a></span>
</li> <li> <span style="white-space: nowrap;"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Ponting/WebHome"><img src="TheAnnotator_files/web-bg.gif" alt="" style="background-color: rgb(204, 255, 204);" border="0" height="16" width="16">&nbsp;Ponting</a></span>
</li> <li> <span style="white-space: nowrap;"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/PostDocs/WebHome"><img src="TheAnnotator_files/web-bg.gif" alt="" style="background-color: rgb(208, 208, 208);" border="0" height="16" width="16">&nbsp;PostDocs</a></span>
</li> <li> <span style="white-space: nowrap;"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/PreDocs/WebHome"><img src="TheAnnotator_files/web-bg.gif" alt="" style="background-color: rgb(208, 208, 208);" border="0" height="16" width="16">&nbsp;PreDocs</a></span>
</li> <li> <span style="white-space: nowrap;"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/Sandbox/WebHome"><img src="TheAnnotator_files/web-bg.gif" alt="" style="background-color: rgb(185, 218, 255);" border="0" height="16" width="16">&nbsp;Sandbox</a></span>
</li> <li> <span style="white-space: nowrap;"><a href="https://wiki.anat.ox.ac.uk/twiki/bin/view/TWiki/WebHome"><img src="TheAnnotator_files/web-bg.gif" alt="" style="background-color: rgb(255, 216, 170);" border="0" height="16" width="16">&nbsp;TWiki</a></span>
</li></ul> 
</div><!-- /patternLeftBarContents--></div><!-- /patternLeftBar-->
</div><!-- /patternFloatWrap-->
<div class="clear">&nbsp;</div>
</div><!-- /patternOuter--></div><!-- /patternWrapper--><div id="patternTopBar"><div id="patternTopBarContents"><table style="width: 100%; margin-top: 11px;" border="0" cellpadding="0" cellspacing="0">
<tbody><tr><td valign="middle"><a href="http://twiki.org/"><img src="TheAnnotator_files/T-logo-140x40-t.gif" alt="Powered by TWiki" border="0"></a></td>
<td class="patternMetaMenu" align="right" valign="top">
 <ul>
<li> <form name="jumpForm" action="/twiki/bin/view/Ponting/TheAnnotator"><input id="jumpFormField" class="twikiInputField patternFormFieldDefaultColor" name="topic" size="14" value="Jump" onfocus="clearDefaultandCSS(this);" onblur="setDefaultText(this);" type="text"></form>
</li> <li> <form name="quickSearchForm" action="/twiki/bin/view/Ponting/WebSearch"><input class="twikiInputField patternFormFieldDefaultColor" name="search" value="Search" size="14" onfocus="clearDefaultandCSS(this);" onblur="setDefaultText(this);" type="text"></form>
</li> <li> 
</li></ul> 
</td></tr></tbody></table></div></div><!-- /patternTopBar--><div id="patternBottomBar"><div id="patternBottomBarContents"><span class="twikiRight"> <a href="http://twiki.org/"><img src="TheAnnotator_files/T-logo-80x15.gif" alt="This site is powered by the TWiki collaboration platform" title="This site is powered by the TWiki collaboration platform" border="0" height="15" width="80"></a></span>Copyright
© by the contributing authors. All material on this collaboration
platform is the property of the contributing authors. <br> Ideas, requests, problems regarding TWiki? <a href="mailto:andreas.heger@dpag.ox.ac.uk?subject=TWiki%20Feedback%20on%20Ponting.TheAnnotator">Send feedback</a> 
</div><!-- /patternBottomBarContents--></div><!-- /patternBottomBar-->
</div><!-- /patternPage-->
</div><!-- /patternPageShadow-->
</div><!-- /patternScreen-->
</body></html>